src/vos/ilog.c

```diff
@@ -1552,3 +1392,257 @@ ilog_fetch_finish(struct ilog_entries *entries)
 	if (!daos_handle_is_inval(priv->ip_ih))
 		dbtree_iter_finish(priv->ip_ih);
 }
+
+static int
+remove_ilog_entry(struct ilog_context *lctx, daos_handle_t *toh,
+		  const struct ilog_entry *entry, int *removed)
+{
+	struct ilog_id	id = entry->ie_id;
+	d_iov_t		iov;
+	int		rc;
+
+	rc = ilog_tx_begin(lctx);
+	if (rc != 0)
+		return rc;
+	D_DEBUG(DB_IO, "Removing ilog entry at "DF_U64"\n",
+		entry->ie_id.id_epoch);
+	d_iov_set(&iov, &id, sizeof(id));
+	rc = dbtree_delete(*toh, BTR_PROBE_EQ, &iov, lctx);
+	if (rc != 0) {
+		D_ERROR("Could not remove entry from tree: "DF_RC"\n",
+			DP_RC(rc));
+		return rc;
+	}
+	D_DEBUG(DB_IO, "Removed ilog entry at "DF_U64"\n",
+		entry->ie_id.id_epoch);
+
+	(*removed)++;
+
+	return 0;
+}
+
+struct agg_arg {
+	const daos_epoch_range_t	*aa_epr;
+	const struct ilog_entry		*aa_prev;
+	const struct ilog_entry		*aa_prior_punch;
+	daos_epoch_t			 aa_punched;
+	bool				 aa_discard;
+};
+
+enum {
+	AGG_RC_DONE,
+	AGG_RC_NEXT,
+	AGG_RC_REMOVE,
+	AGG_RC_REMOVE_PREV,
+};
+
+static int
+check_agg_entry(const struct ilog_entry *entry, struct agg_arg *agg_arg)
+{
+	int	rc;
+
+	D_DEBUG(DB_TRACE, "Entry "DF_U64" punch=%s prev="DF_U64
+		" prior_punch="DF_U64"\n", entry->ie_id.id_epoch,
+		entry->ie_punch ? "yes" : "no",
+		agg_arg->aa_prev ? agg_arg->aa_prev->ie_id.id_epoch : 0,
+		agg_arg->aa_prior_punch ?
+		agg_arg->aa_prior_punch->ie_id.id_epoch : 0);
+
+	if (entry->ie_id.id_epoch > agg_arg->aa_epr->epr_hi)
+		D_GOTO(done, rc = AGG_RC_DONE);
+	if (entry->ie_id.id_epoch < agg_arg->aa_epr->epr_lo) {
+		if (entry->ie_id.id_epoch <= agg_arg->aa_punched) {
```

Comment:
I was thinking that "epr_lo <= aa_punched <= epr_hi", so this check looks redundant to me, did I miss anything?

Reply:
If the parent isn’t punched, aa_punched will be 0